official/coin_build/man/man3/SoShapeKit.3
.TH "SoShapeKit" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SoShapeKit \- The \fBSoShapeKit\fP class provides templates to insert what is usually needed for shape nodes\&.
.PP
Node kit structure (new entries versus parent class marked with arrow prefix):  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/nodekits/SoShapeKit\&.h>\fP
.PP
Inherits \fBSoSeparatorKit\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const"
.br
.RI "Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. "
.ti -1c
.RI "virtual const \fBSoNodekitCatalog\fP * \fBgetNodekitCatalog\fP (void) const"
.br
.ti -1c
.RI "\fBSoShapeKit\fP (void)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.ti -1c
.RI "static const \fBSoNodekitCatalog\fP * \fBgetClassNodekitCatalog\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "virtual const \fBSoFieldData\fP * \fBgetFieldData\fP (void) const"
.br
.ti -1c
.RI "virtual \fB~SoShapeKit\fP ()"
.br
.ti -1c
.RI "virtual void \fBsetDefaultOnNonWritingFields\fP (void)"
.br
.in -1c
.SS "Static Protected Member Functions"

.in +1c
.ti -1c
.RI "static const \fBSoFieldData\fP ** \fBgetFieldDataPtr\fP (void)"
.br
.ti -1c
.RI "static const \fBSoNodekitCatalog\fP ** \fBgetClassNodekitCatalogPtr\fP (void)"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "\fBSoSFNode\fP \fBcoordinate3\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBcoordinate4\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBlocalTransform\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBmaterialBinding\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBnormal\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBnormalBinding\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBprofileCoordinate2\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBprofileCoordinate3\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBprofileList\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBshape\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBshapeHints\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBshapeSeparator\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBtextureCoordinate2\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBtextureCoordinateBinding\fP"
.br
.ti -1c
.RI "\fBSoSFNode\fP \fBtextureCoordinateFunction\fP"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
The \fBSoShapeKit\fP class provides templates to insert what is usually needed for shape nodes\&.
.PP
Node kit structure (new entries versus parent class marked with arrow prefix): 


.PP
.nf
CLASS SoShapeKit
-->"this"
      "callbackList"
      "topSeparator"
         "pickStyle"
         "appearance"
         "units"
         "transform"
         "texture2Transform"
-->      "materialBinding"
-->      "normalBinding"
-->      "textureCoordinateBinding"
-->      "shapeHints"
-->      "coordinate3"
-->      "coordinate4"
-->      "normal"
-->      "textureCoordinate2"
-->      "profileCoordinate2"
-->      "profileCoordinate3"
-->      "profileList"
         "childList"
-->      "textureCoordinateFunction"
-->      "localTransform"
-->      "shapeSeparator"
-->         "shape"

.fi
.PP
.PP
(See \fBSoBaseKit::printDiagram()\fP for information about the output formatting\&.)
.PP
Detailed information on catalog parts:
.PP
.PP
.nf
CLASS SoShapeKit
PVT   "this",  SoShapeKit  --- 
      "callbackList",  SoNodeKitListPart [ SoCallback, SoEventCallback ] 
PVT   "topSeparator",  SoSeparator  --- 
      "pickStyle",  SoPickStyle  --- 
      "appearance",  SoAppearanceKit  --- 
      "units",  SoUnits  --- 
      "transform",  SoTransform  --- 
      "texture2Transform",  SoTexture2Transform  --- 
      "materialBinding",  SoMaterialBinding  --- 
      "normalBinding",  SoNormalBinding  --- 
      "textureCoordinateBinding",  SoTextureCoordinateBinding  --- 
      "shapeHints",  SoShapeHints  --- 
      "coordinate3",  SoCoordinate3  --- 
      "coordinate4",  SoCoordinate4  --- 
      "normal",  SoNormal  --- 
      "textureCoordinate2",  SoTextureCoordinate2  --- 
      "profileCoordinate2",  SoProfileCoordinate2  --- 
      "profileCoordinate3",  SoProfileCoordinate3  --- 
      "profileList",  SoNodeKitListPart [ SoProfile ] 
      "childList",  SoNodeKitListPart [ SoShapeKit, SoSeparatorKit ] 
      "textureCoordinateFunction",  SoTextureCoordinateFunction  --- , (default type = SoTextureCoordinateDefault)
      "localTransform",  SoTransform  --- 
PVT   "shapeSeparator",  SoSeparator  --- 
      "shape",  SoShape  --- , (default type = SoCube)
.fi
.PP
.PP
(See \fBSoBaseKit::printTable()\fP for information about the output formatting\&.) 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SoShapeKit::SoShapeKit (void)"
Constructor\&. 
.SS "SoShapeKit::~SoShapeKit ()\fC [protected]\fP, \fC [virtual]\fP"
Destructor\&. 
.SH "Member Function Documentation"
.PP 
.SS "\fBSoType\fP SoShapeKit::getTypeId (void) const\fC [virtual]\fP"

.PP
Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. Usage example:
.PP
.PP
.nf
void foo(SoNode * node)
{
  if (node->getTypeId() == SoFile::getClassTypeId()) {
    SoFile * filenode = (SoFile *)node;  // safe downward cast, knows the type
  }
}
.fi
.PP
.PP
For application programmers wanting to extend the library with new nodes, engines, nodekits, draggers or others: this method needs to be overridden in \fIall\fP subclasses\&. This is typically done as part of setting up the full type system for extension classes, which is usually accomplished by using the pre-defined macros available through for instance \fBInventor/nodes/SoSubNode\&.h\fP (SO_NODE_INIT_CLASS and SO_NODE_CONSTRUCTOR for node classes), \fBInventor/engines/SoSubEngine\&.h\fP (for engine classes) and so on\&.
.PP
For more information on writing Coin extensions, see the class documentation of the toplevel superclasses for the various class groups\&. 
.PP
Reimplemented from \fBSoSeparatorKit\fP\&.
.SS "const \fBSoFieldData\fP * SoShapeKit::getFieldData (void) const\fC [protected]\fP, \fC [virtual]\fP"
Returns a pointer to the class-wide field data storage object for this instance\&. If no fields are present, returns \fCNULL\fP\&. 
.PP
Reimplemented from \fBSoSeparatorKit\fP\&.
.SS "const \fBSoNodekitCatalog\fP * SoShapeKit::getNodekitCatalog (void) const\fC [virtual]\fP"
Returns the nodekit catalog which defines the layout of this class' kit\&. 
.PP
Reimplemented from \fBSoSeparatorKit\fP\&.
.SS "void SoShapeKit::setDefaultOnNonWritingFields (void)\fC [protected]\fP, \fC [virtual]\fP"
(Be aware that this method is unlikely to be of interest to the application programmer who does not want to extend the library with new custom nodekits or draggers\&. If you indeed \fIare\fP writing extensions, see the information in the \fBSoBaseKit\fP class documentation\&.)
.PP
This is a virtual method, and the code in it should call \fBSoField::setDefault()\fP with argument \fCTRUE\fP on part fields that should not be written upon scenegraph export operations\&.
.PP
This is typically done when:
.PP
.PD 0
.IP "1." 4
field value is \fCNULL\fP and part is \fCNULL\fP by default 
.PP

.IP "2." 4
it is a leaf \fBSoGroup\fP or \fBSoSeparator\fP node with no children 
.PP

.IP "3." 4
it is a leaf listpart with no children and an \fBSoGroup\fP or \fBSoSeparator\fP container 
.PP

.IP "4." 4
it is a non-leaf part and it's of \fBSoGroup\fP type and all fields are at their default values 
.PP

.PP
.PP
Subclasses should usually override this to do additional settings for new member fields\&. From the subclass, do remember to call 'upwards' to your superclass' \fBsetDefaultOnNonWritingFields()\fP method\&. 
.PP
Reimplemented from \fBSoSeparatorKit\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


coin4/src/nodekits/SoShapeKit.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif // HAVE_CONFIG_H

#ifdef HAVE_NODEKITS

/*!
  \class SoShapeKit SoShapeKit.h Inventor/nodekits/SoShapeKit.h
  \brief The SoShapeKit class provides templates to insert what is usually needed for shape nodes.
  \ingroup nodekits

  \NODEKIT_PRE_DIAGRAM

  \verbatim
  CLASS SoShapeKit
  -->"this"
        "callbackList"
        "topSeparator"
           "pickStyle"
           "appearance"
           "units"
           "transform"
           "texture2Transform"
  -->      "materialBinding"
  -->      "normalBinding"
  -->      "textureCoordinateBinding"
  -->      "shapeHints"
  -->      "coordinate3"
  -->      "coordinate4"
  -->      "normal"
  -->      "textureCoordinate2"
  -->      "profileCoordinate2"
  -->      "profileCoordinate3"
  -->      "profileList"
           "childList"
  -->      "textureCoordinateFunction"
  -->      "localTransform"
  -->      "shapeSeparator"
  -->         "shape"
  \endverbatim

  \NODEKIT_POST_DIAGRAM


  \NODEKIT_PRE_TABLE

  \verbatim
  CLASS SoShapeKit
  PVT   "this",  SoShapeKit  --- 
        "callbackList",  SoNodeKitListPart [ SoCallback, SoEventCallback ] 
  PVT   "topSeparator",  SoSeparator  --- 
        "pickStyle",  SoPickStyle  --- 
        "appearance",  SoAppearanceKit  --- 
        "units",  SoUnits  --- 
        "transform",  SoTransform  --- 
        "texture2Transform",  SoTexture2Transform  --- 
        "materialBinding",  SoMaterialBinding  --- 
        "normalBinding",  SoNormalBinding  --- 
        "textureCoordinateBinding",  SoTextureCoordinateBinding  --- 
        "shapeHints",  SoShapeHints  --- 
        "coordinate3",  SoCoordinate3  --- 
        "coordinate4",  SoCoordinate4  --- 
        "normal",  SoNormal  --- 
        "textureCoordinate2",  SoTextureCoordinate2  --- 
        "profileCoordinate2",  SoProfileCoordinate2  --- 
        "profileCoordinate3",  SoProfileCoordinate3  --- 
        "profileList",  SoNodeKitListPart [ SoProfile ] 
        "childList",  SoNodeKitListPart [ SoShapeKit, SoSeparatorKit ] 
        "textureCoordinateFunction",  SoTextureCoordinateFunction  --- , (default type = SoTextureCoordinateDefault)
        "localTransform",  SoTransform  --- 
  PVT   "shapeSeparator",  SoSeparator  --- 
        "shape",  SoShape  --- , (default type = SoCube)
  \endverbatim

  \NODEKIT_POST_TABLE
*/

#include <Inventor/nodekits/SoShapeKit.h>

#include <Inventor/nodekits/SoNodeKitListPart.h>
#include <Inventor/nodes/SoMaterialBinding.h>
#include <Inventor/nodes/SoNormalBinding.h>
#include <Inventor/nodes/SoTextureCoordinateBinding.h>
#include <Inventor/nodes/SoShapeHints.h>
#include <Inventor/nodes/SoCoordinate3.h>
#include <Inventor/nodes/SoCoordinate4.h>
#include <Inventor/nodes/SoNormal.h>
#include <Inventor/nodes/SoTextureCoordinate2.h>
#include <Inventor/nodes/SoProfileCoordinate2.h>
#include <Inventor/nodes/SoProfileCoordinate3.h>
#include <Inventor/nodes/SoTextureCoordinateDefault.h>
#include <Inventor/nodes/SoTransform.h>
#include <Inventor/nodes/SoSeparator.h>
#include <Inventor/nodes/SoCube.h>
#include <Inventor/nodes/SoProfile.h>

#include "nodekits/SoSubKitP.h"


SO_KIT_SOURCE(SoShapeKit);


/*!
  Constructor.
*/
SoShapeKit::SoShapeKit(void)
{
  SO_KIT_INTERNAL_CONSTRUCTOR(SoShapeKit);

  // Note: we must use "" instead of , , to humour MS VisualC++ 6.

  SO_KIT_ADD_CATALOG_ENTRY(materialBinding, SoMaterialBinding, true, topSeparator, normalBinding, true);
  SO_KIT_ADD_CATALOG_ENTRY(normalBinding, SoNormalBinding, true, topSeparator, textureCoordinateBinding, true);
  SO_KIT_ADD_CATALOG_ENTRY(textureCoordinateBinding, SoTextureCoordinateBinding, true, topSeparator, shapeHints, true);
  SO_KIT_ADD_CATALOG_ENTRY(shapeHints, SoShapeHints, true, topSeparator, coordinate3, true);
  SO_KIT_ADD_CATALOG_ENTRY(coordinate3, SoCoordinate3, true, topSeparator, coordinate4, true);
  SO_KIT_ADD_CATALOG_ENTRY(coordinate4, SoCoordinate4, true, topSeparator, normal, true);
  SO_KIT_ADD_CATALOG_ENTRY(normal, SoNormal, true, topSeparator, textureCoordinate2, true);
  SO_KIT_ADD_CATALOG_ENTRY(textureCoordinate2, SoTextureCoordinate2, true, topSeparator, profileCoordinate2, true);
  SO_KIT_ADD_CATALOG_ABSTRACT_ENTRY(textureCoordinateFunction, SoTextureCoordinateFunction, SoTextureCoordinateDefault, true, topSeparator, localTransform, true);
  SO_KIT_ADD_CATALOG_ENTRY(profileCoordinate2, SoProfileCoordinate2, true, topSeparator, profileCoordinate3, true);
  SO_KIT_ADD_CATALOG_ENTRY(profileCoordinate3, SoProfileCoordinate3, true, topSeparator, profileList, true);
  SO_KIT_ADD_CATALOG_LIST_ENTRY(profileList, SoGroup, true, topSeparator, childList, SoProfile, true);
  SO_KIT_ADD_CATALOG_ENTRY(localTransform, SoTransform, true, topSeparator, shapeSeparator, true);
  SO_KIT_ADD_CATALOG_ENTRY(shapeSeparator, SoSeparator, true, topSeparator, "", false);
  SO_KIT_ADD_CATALOG_ABSTRACT_ENTRY(shape, SoShape, SoCube, false, shapeSeparator, "", true);

  SO_KIT_INIT_INSTANCE();
}

/*!
  Destructor.
*/
SoShapeKit::~SoShapeKit()
{
}

// Documented in superclass.
void
SoShapeKit::initClass(void)
{
  SO_KIT_INTERNAL_INIT_CLASS(SoShapeKit, SO_FROM_INVENTOR_1);
}

// Documented in superclass.
void
SoShapeKit::setDefaultOnNonWritingFields(void)
{
  this->shapeSeparator.setDefault(true);
  inherited::setDefaultOnNonWritingFields();
}

#endif // HAVE_NODEKITS
